File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.java
import java.util.*;
class Solution {
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        
        Map<String, String> parent = new HashMap<>();
        Map<String, Integer> res = new HashMap<>();
        for(int i=0; i<enroll.length; i++) {
            parent.put(enroll[i], referral[i]); //판매원의 부모를 기록
            res.put(enroll[i], 0);
        }
        
    
        for(int i=0; i<seller.length; i++) {
            String cur = seller[i];
            int price = amount[i] * 100;
            while(true) {
                if(cur.equals("-")) break;
                
                int remain = (int) (price * 0.1);
                if(remain == 0) {
                    res.put(cur, res.get(cur) + price);
                    break;
                }
                else { //10프로 이월이 된다면
                    res.put(cur, res.get(cur) + (price-remain));
                    price = remain;
                    cur = parent.get(cur);
                }
            }
        }
        
        int[] ans = new int[enroll.length];
        for(int i=0; i<enroll.length; i++) {
            ans[i] = res.get(enroll[i]);
        }
        
        return ans;
    }
}

File: 프로그래머스/3/77486. 다단계 칫솔 판매/다단계 칫솔 판매.js
function solution(enroll, referral, seller, amount) {
    // 이름 - 번호
    const referralMap = {}
    // 판매자 - 추천인
    const recommendMap = {}
    
    const N = enroll.length
    const money = Array(N).fill(0)
    
    for (let i=0; i<N; i++) {
        referralMap[enroll[i]] = i
        recommendMap[enroll[i]] = referral[i]
    }
    
    for (let i=0; i<seller.length; i++) {
        
        const sellerWho = seller[i]
        const sellerMoney = amount[i] * 100
        
        let curWho = sellerWho
        let curMoney = sellerMoney
        
        while (curMoney >= 1 && curWho != "-") {
            
            const idx = referralMap[curWho]
            // 본인이 가져갈 돈
            
            const recommed = recommendMap[curWho] // 추천해준사람
            
            if (Math.floor(curMoney * 0.1) === 0 ) {
                money[idx] += curMoney;
                break
            }
            
            const remain = Math.floor(curMoney * 0.1)
            const ownMoney = curMoney - remain
            money[idx] += ownMoney
            curMoney = remain
            curWho = recommed
            
        }
    }
    
    return money
}